cml-core/src/core/inference.rs

```diff
@@ -24,13 +24,13 @@ pub trait Inference<M, F, T, C: Manager> {
 
     async fn inference<FN>(
         &self,
-        metadata: Metadata<F>,
+        metadata: &Metadata<F>,
         available_status: &[&str],
-        data: &mut Vec<NewSample<F>>,
+        data: Vec<NewSample<F>>,
         batch_state: &SharedBatchState,
         pool: &Pool<C>,
         inference_fn: FN,
     ) -> Result<()>
     where
-        FN: FnOnce(&mut Vec<NewSample<F>>, &str, T) -> Vec<NewSample<F>>;
+        FN: FnOnce(Vec<NewSample<F>>, &str, T) -> Vec<NewSample<F>>;
```

Comment:
You should use `&mut [T]` here and implicitly return an unit type (`()`). It's worth noting that we aim to avoid any unnecessary memory allocation, which is why we use `FnOnce`.